File: C++/559. Maximum Depth of N-ary Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

// class Solution {
// public:
//     int maxDepth(Node* root) {
//         if (root == NULL) return 0;
//         int max_ = 0;
//         for (auto child: root->children){
//             max_ = max(max_, maxDepth(child));
//         }
//         return 1 + max_;
//     }
// };

class Solution {
public:
    int maxDepth(Node* root) {
        int depth = 0;
        if (root == NULL) return depth;
        queue<Node*> que;
        que.push(root);
        while (!que.empty()) {
            depth++;
            int size = que.size();
            for (int i = 0; i < size; i++) {
                Node* cur = que.front(); que.pop();
                for (auto child : cur->children) {
                    que.push(child);
                }
            }
        }
        return depth;
    }
};

File: Python/559. Maximum Depth of N-ary Tree.py
'''
1.递归
2.迭代
'''
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


# class Solution:
#     def maxDepth(self, root: 'Node') -> int:
#         if not root:
#             return 0
#         max_ = 0
#         for child in root.children:
#             max_ = max(max_, self.maxDepth(child))
#         return 1 + max_

class Solution:
    def maxDepth(self, root: 'Node') -> int:
        depth = 0
        if not root:
            return depth
        que = [root]
        while que:
            depth += 1
            size = len(que)
            for _ in range(size):
                cur = que.pop(0)
                for child in cur.children:
                    que.append(child)
        return depth